include/DeterministicNoiseSimulator.hpp

```diff
@@ -0,0 +1,120 @@
+    //                {"final_fidelity", std::to_string(final_fidelity)},
+    //        };
+    //    };
+
+    [[nodiscard]] std::string intToString(long target_number, char value) const;
+
+    void apply_det_noise_sequential(const qc::Targets& targets);
+
+    std::map<std::string, double> AnalyseState(int nr_qubits);
+
+    void setNoiseEffects(const std::string& cGateNoise) { gate_noise_types = cGateNoise; }
+
+    [[nodiscard]] dd::QubitCount getNumberOfQubits() const override { return qc->getNqubits(); };
+
+    [[nodiscard]] std::size_t getNumberOfOps() const override { return qc->getNops(); };
+
+    [[nodiscard]] std::string getName() const override { return qc->getName(); };
+
+    const std::map<char, int> noise_effects = {
+            {'B', 2}, //Bit-flip
+            {'P', 2}, //Phase-flip
+            {'A', 2}, //Amplitude Damping
+            {'D', 4}, //Depolarisation
+    };
+
+    double           noise_probability = 0.0;
+    dd::ComplexValue sqrt_amplitude_damping_probability{};
+    dd::ComplexValue one_minus_sqrt_amplitude_damping_probability{};
+
+    dd::Package::mEdge density_root_edge;
+
+    bool weird_value_i_dont_understand_but_something_with_sequential = false;
+    char MeasureOneCollapsing(dd::Qubit index);
+
+private:
+    static constexpr unsigned short NoiseSLOTS = 16384;
+    static constexpr unsigned short NoiseMASK  = NoiseSLOTS - 1;
+    struct NoiseEntry {
+        dd::Package::mNode *   a, *r; // a is the argument, r is the result
+        dd::ComplexValue       aw, rw;
+        dd::NoiseOperationKind which; // type of operation
+        short                  line[std::numeric_limits<dd::Qubit>::max() + 1];
+    };
+
+    std::unique_ptr<qc::QuantumComputation>& qc;
+
+    std::string gate_noise_types;
+
+    std::array<NoiseEntry, NoiseSLOTS> NoiseTable{};
+
+    //    const unsigned int step_number;
+    //    const double       step_fidelity;
+    //    unsigned long long approximation_runs{0};
+    //    long double        final_fidelity{1.0L};
+
+    void ApplyAmplitudeDampingToNode(std::array<dd::Package::mEdge, 4>& e);
+
+    void ApplyPhaseFlipToNode(std::array<dd::Package::mEdge, 4>& e);
+
+    void ApplyDepolaritationToNode(std::array<dd::Package::mEdge, 4>& e);
+
+    void generate_gate(dd::Package::mEdge* pointer_for_matrices, char noise_type, dd::Qubit target);
+
+    dd::Package::mEdge makeZeroDensityOperator(dd::QubitCount n);
+
+    dd::Package::mEdge ApplyNoiseEffects(dd::Package::mEdge density_op, const std::unique_ptr<qc::Operation>& op, unsigned char maxDepth);
+    unsigned long      noiseHash(dd::Package::mNode* a, const dd::ComplexValue& aw, const short* line, unsigned short nQubits);
+    dd::Package::mEdge noiseLookup(const dd::Package::mEdge& a, const short* line, unsigned short nQubits);
+    void               noiseInsert(const dd::Package::mEdge& a, const short* line, const dd::Package::mEdge& r, unsigned short nQubits);
```

Comment:
This should be covered by using the `NoiseOperationTable` as well

Reply:
See above